### PR TITLE
#6975 lib/ui TS migration - 1

### PR DESCRIPTION
**Pull request by leoyli**
Issue: #6975 (do not close)

## What I did

- Move Sidebar components in the lib/ui into TypeScript

## How to test

- Let CI to run test to see if regressions happened 

related #6975.

---

Along the way to make this PR, I made several `FIXME` comments hope someday someone or myself can pick these up. These changes is laying out of the domain of this package so I did not cleanup these for the moment, otherwise I feel there are endless digging and really overload what I really want to finish in this issue.

My personal experiences/feeling on TypeScript is very positive, but I do feel we need more love on making things really typed, not just looks typed...  (don't want just to abuse `any`...) And I know this can be hard to do, so virtually never going to be the major focus I guess, that is why I need to specifically bring this to the table.

Finally, I found `emotion` brings too many different paradigms on writing CSS-in-JS, so some typing is always not "right" for many reasons. Maybe we could consider to migrate (back) to `styled-components` (and use `stylelint` get css checked)?  This could also make the storybook design system more easier to be integrated in the future.

### CODE REVIEW

File: lib/ui/src/components/sidebar/SidebarItem.tsx [hunk 1]
```diff
@@ -96,15 +105,22 @@ export const Item = styled(({ className, children, id }) => (
...
+  name = 'loading story',
+  isComponent = false,
+  isLeaf = false,
+  isExpanded = false,
+  isSelected = false,
   ...props
-}) {
-  let iconName;
+}: SidebarItemProps) => {
+  let iconName: React.ComponentProps<typeof Icons>['icon'];
```
Comment (shilman) [inline lib/ui/src/components/sidebar/SidebarItem.tsx]: Is this the right way to do this? Wondering if it's better to export a type? This looks a little verbose to me, but I'm a typescript newb.

Comment (leoyli) [reply lib/ui/src/components/sidebar/SidebarItem.tsx]: The top-level `@storybook/components` isn't directly expose the `IconsProps`, which makes import path a bit hard to find (need to dig where the type is defined in the `node_module`)... So I just use the helper type from React to extract the `IconsProps` (`IconsProps === React.ComponentProps<typeof Icons>`).

**Comment by leoyli** [comment]
The typing on emotion must be somewhere really wrong.  

For example:

```ts
const FilterField = styled.input(({ theme }) => ...
```

The above theoretically should have nice tying without passing typed arguments.  But in this case, when the component being used, it is always asking for the missing `theme` props... 🤷‍♂ 

To fix this, then we ended up getting something like:

```ts

type FilterFieldProps = React.ComponentProps<'input'>; // or something equivalent 

const FilterField = styled.input<FilterFieldProps>(({ theme }) => ...
```

This is really annoying... 😭

**Comment by leoyli** [comment]
@shilman, @ndelangen,

My build failed due to module `@storybook/react` not found by `SidebarSubheading.story.ts` in CI env, should I added it into the devDependencies?

File: lib/ui/src/components/sidebar/SidebarSubheading.stories.tsx [hunk 2]
```diff
@@ -1,10 +1,10 @@
...
 
 import SidebarSubheading from './SidebarSubheading';
 
-storiesOf('UI|Sidebar/SidebarSubheading', module)
-  .addParameters({
-    component: SidebarSubheading,
-  })
-  .add('default', () => <SidebarSubheading>Subheading</SidebarSubheading>);
+export default {
+  Component: SidebarSubheading,
```
Comment (shilman) [inline lib/ui/src/components/sidebar/SidebarSubheading.stories.tsx]: Component => component

Comment (ndelangen) [reply lib/ui/src/components/sidebar/SidebarSubheading.stories.tsx]: there's loads of other components stories that have `Component`?

Comment (shilman) [reply lib/ui/src/components/sidebar/SidebarSubheading.stories.tsx]: Well then they should probably be fixed 😂

Comment (ndelangen) [reply lib/ui/src/components/sidebar/SidebarSubheading.stories.tsx]: ok

**Review by ndelangen — APPROVED** [review]

**Comment by leoyli** [comment]
@ndelangen, oh thanks for taking care of this ❤️

File: lib/ui/package.json [hunk 3]
```diff
@@ -23,6 +23,8 @@
   },
   "dependencies": {
     "@storybook/addons": "5.2.0-beta.19",
+    "@storybook/addon-actions": "5.2.0-beta.19",
```
Comment (stof) [inline lib/ui/package.json]: why is `@storybook/ui` now installing some addons ? Shouldn't they be dev dependencies instead ?

Comment (ndelangen) [reply lib/ui/package.json]: I'll check it out.

Comment (ndelangen) [reply lib/ui/package.json]: You are correct, I'll issue a PR

**Pull request merged**